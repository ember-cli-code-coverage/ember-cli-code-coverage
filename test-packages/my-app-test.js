'use strict';

const fs = require('fs-extra');
const util = require('util');
const rimraf = util.promisify(require('rimraf'));
const { dir, file } = require('chai-files');
const path = require('path');
const execa = require('execa');

const BASE_PATH = path.join(__dirname, 'my-app');

describe('app coverage generation', function () {
  jest.setTimeout(10000000);

  beforeEach(async function () {
    await rimraf(`${BASE_PATH}/coverage*`);
    await execa('git', ['clean', '-f', 'my-app'], { cwd: __dirname });
    await execa('git', ['restore', 'my-app'], { cwd: __dirname });
  });

  afterEach(async function () {
    await rimraf(`${BASE_PATH}/coverage*`);
    await execa('git', ['clean', '-f', 'my-app'], { cwd: __dirname });
    await execa('git', ['restore', 'my-app'], { cwd: __dirname });
  });

  it('runs coverage when env var is set', async function () {
    dir(`${BASE_PATH}/coverage`).assertDoesNotExist();

    let env = { COVERAGE: 'true' };
    await execa('ember', ['test'], { cwd: BASE_PATH, env });
    file(`${BASE_PATH}/coverage/lcov-report/index.html`).assertIsNotEmpty();
    file(`${BASE_PATH}/coverage/index.html`).assertIsNotEmpty();
    let summary = fs.readJSONSync(`${BASE_PATH}/coverage/coverage-summary.json`);
    expect(summary.total.lines.pct).to.equal(66.67);
  });

  it('does not run coverage when env var is NOT set', async function () {
    dir(`${BASE_PATH}/coverage`).assertDoesNotExist();

    let env = { COVERAGE: 'false' };
    await execa('ember', ['test'], { cwd: BASE_PATH, env });
    dir(`${BASE_PATH}/coverage`).assertDoesNotExist();
  });

  it('excludes files when the configuration is set', async function () {
    fs.copySync(`${BASE_PATH}/config/-coverage-excludes.js`, `${BASE_PATH}/config/coverage.js`);

    let env = { COVERAGE: 'true' };
    await execa('ember', ['test'], { cwd: BASE_PATH, env });
    file(`${BASE_PATH}/coverage/lcov-report/index.html`).assertIsNotEmpty();
    file(`${BASE_PATH}/coverage/index.html`).assertIsNotEmpty();
    let summary = fs.readJSONSync(`${BASE_PATH}/coverage/coverage-summary.json`);
    expect(summary.total.lines.pct).to.equal(75);
  });

  it('merges coverage when tests are run in parallel', async function () {
    dir(`${BASE_PATH}/coverage`).assertDoesNotExist();

    let env = { COVERAGE: 'true' };
    await execa('ember', ['exam', '--split=2', '--parallel=true'], { cwd: BASE_PATH, env });
    file(`${BASE_PATH}/coverage/lcov-report/index.html`).assertIsNotEmpty();
    file(`${BASE_PATH}/coverage/index.html`).assertIsNotEmpty();
    let summary = fs.readJSONSync(`${BASE_PATH}/coverage/coverage-summary.json`);
    expect(summary.total.lines.pct).to.equal(66.67);
  });

  it('uses parallel configuration and merges coverage when merge-coverage command is issued', async function () {
    dir(`${BASE_PATH}/coverage`).assertDoesNotExist();
    fs.copySync(`${BASE_PATH}/config/-coverage-parallel.js`, `${BASE_PATH}/config/coverage.js`);

    let env = { COVERAGE: 'true' };
    await execa('ember', ['exam', '--split=2', '--parallel=true'], { cwd: BASE_PATH, env });
    dir(`${BASE_PATH}/coverage`).assertDoesNotExist();

    await execa('ember', ['coverage-merge'], { cwd: BASE_PATH });
    file(`${BASE_PATH}/coverage/lcov-report/index.html`).assertIsNotEmpty();
    file(`${BASE_PATH}/coverage/index.html`).assertIsNotEmpty();
    let summary = fs.readJSONSync(`${BASE_PATH}/coverage/coverage-summary.json`);
    expect(summary.total.lines.pct).to.equal(66.67);
  });

  it('uses nested coverageFolder and parallel configuration and run merge-coverage', async function () {
    let coverageFolder = `${BASE_PATH}/coverage/abc/easy-as/123`;

    dir(coverageFolder).assertDoesNotExist();
    fs.copySync(
      `${BASE_PATH}/config/-coverage-nested-folder.js`,
      `${BASE_PATH}/config/coverage.js`
    );

    let env = { COVERAGE: 'true' };
    await execa('ember', ['exam', '--split=2', '--parallel=true'], { cwd: BASE_PATH, env });
    dir(coverageFolder).assertDoesNotExist();

    await execa('ember', ['coverage-merge'], { cwd: BASE_PATH });
    file(`${coverageFolder}/lcov-report/index.html`).assertIsNotEmpty();
    file(`${coverageFolder}/index.html`).assertIsNotEmpty();
    let summary = fs.readJSONSync(`${coverageFolder}/coverage-summary.json`);
    expect(summary.total.lines.pct).to.equal(66.67);
  });

  it('runs coverage when a module has an import error', async function () {
    dir(`${BASE_PATH}/coverage`).assertDoesNotExist();
    fs.copySync(`${BASE_PATH}/-error-module.js`, `${BASE_PATH}/app/error-module.js`);

    let env = { COVERAGE: 'true' };
    await execa('ember', ['test'], { cwd: BASE_PATH, env });
    dir(`${BASE_PATH}/coverage`).assertExists();
  });
});
