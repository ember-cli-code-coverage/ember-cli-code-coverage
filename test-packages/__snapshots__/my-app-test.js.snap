// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app coverage generation excludes files when the configuration is set 1`] = `
Object {
  "app/app.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": Object {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/router.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/utils/my-covered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "app/utils/my-uncovered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
    "statements": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 0,
    },
  },
  "total": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 50,
      "skipped": 0,
      "total": 2,
    },
    "lines": Object {
      "covered": 6,
      "pct": 75,
      "skipped": 0,
      "total": 8,
    },
    "statements": Object {
      "covered": 6,
      "pct": 75,
      "skipped": 0,
      "total": 8,
    },
  },
}
`;

exports[`app coverage generation merges coverage when tests are run in parallel 1`] = `
Object {
  "app/app.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": Object {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/router.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/utils/my-covered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "app/utils/my-uncovered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
  },
  "total": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "lines": Object {
      "covered": 6,
      "pct": 66.67,
      "skipped": 0,
      "total": 9,
    },
    "statements": Object {
      "covered": 6,
      "pct": 66.67,
      "skipped": 0,
      "total": 9,
    },
  },
}
`;

exports[`app coverage generation runs coverage when a module has an import error 1`] = `
Object {
  "app/app.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": Object {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/error-module.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "app/router.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/utils/my-covered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "app/utils/my-uncovered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
  },
  "total": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "lines": Object {
      "covered": 7,
      "pct": 70,
      "skipped": 0,
      "total": 10,
    },
    "statements": Object {
      "covered": 7,
      "pct": 70,
      "skipped": 0,
      "total": 10,
    },
  },
}
`;

exports[`app coverage generation runs coverage when env var is set 1`] = `
Object {
  "app/app.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": Object {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/router.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/utils/my-covered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "app/utils/my-uncovered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
  },
  "total": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "lines": Object {
      "covered": 6,
      "pct": 66.67,
      "skipped": 0,
      "total": 9,
    },
    "statements": Object {
      "covered": 6,
      "pct": 66.67,
      "skipped": 0,
      "total": 9,
    },
  },
}
`;

exports[`app coverage generation uses nested coverageFolder and parallel configuration and run merge-coverage 1`] = `
Object {
  "app/app.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": Object {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/router.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/utils/my-covered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "app/utils/my-uncovered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
  },
  "total": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "lines": Object {
      "covered": 6,
      "pct": 66.67,
      "skipped": 0,
      "total": 9,
    },
    "statements": Object {
      "covered": 6,
      "pct": 66.67,
      "skipped": 0,
      "total": 9,
    },
  },
}
`;

exports[`app coverage generation uses parallel configuration and merges coverage when merge-coverage command is issued 1`] = `
Object {
  "app/app.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "lines": Object {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
    "statements": Object {
      "covered": 4,
      "pct": 100,
      "skipped": 0,
      "total": 4,
    },
  },
  "app/router.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "statements": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
  },
  "app/utils/my-covered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 1,
      "pct": 100,
      "skipped": 0,
      "total": 1,
    },
  },
  "app/utils/my-uncovered-util.js": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "lines": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
    "statements": Object {
      "covered": 0,
      "pct": 0,
      "skipped": 0,
      "total": 1,
    },
  },
  "total": Object {
    "branches": Object {
      "covered": 0,
      "pct": 100,
      "skipped": 0,
      "total": 0,
    },
    "functions": Object {
      "covered": 1,
      "pct": 33.33,
      "skipped": 0,
      "total": 3,
    },
    "lines": Object {
      "covered": 6,
      "pct": 66.67,
      "skipped": 0,
      "total": 9,
    },
    "statements": Object {
      "covered": 6,
      "pct": 66.67,
      "skipped": 0,
      "total": 9,
    },
  },
}
`;
